tests/integration/v3_leadership_test.go

```diff
@@ -179,3 +180,77 @@ func TestTransferLeadershipWithLearner(t *testing.T) {
 		t.Error("timed out waiting for leader transition")
 	}
 }
+
+func TestFirstCommitNotification(t *testing.T) {
+	BeforeTest(t)
+
+	clusterSize := 3
+	cluster := NewClusterV3(t, &ClusterConfig{Size: clusterSize})
+	defer cluster.Terminate(t)
+
+	oldLeaderIdx := cluster.WaitLeader(t)
+	oldLeaderClient := cluster.Client(oldLeaderIdx)
+
+	newLeaderIdx := (oldLeaderIdx + 1) % clusterSize
+	newLeaderId := uint64(cluster.Members[newLeaderIdx].ID())
+
+	notifiers := make(map[int]<-chan struct{}, clusterSize)
+	for i, clusterMember := range cluster.Members {
+		notifiers[i] = clusterMember.s.FirstCommitInTermNotify()
+	}
+
+	_, err := oldLeaderClient.MoveLeader(context.Background(), newLeaderId)
+
+	if err != nil {
+		t.Errorf("got error during leadership transfer: %v", err)
+	}
+
+	leaderAppliedIndex := cluster.Members[newLeaderIdx].s.AppliedIndex()
+	doneC := make(chan string, clusterSize)
+
+	for i, notifier := range notifiers {
+		go func(member *member, notifier <-chan struct{}) {
```

Comment:
Please define it as an explicit helper function.  It's pretty long, and "deeply" indented, that harms readability.